sloth/toolBox.py

```diff
@@ -628,5 +632,119 @@ def mapDataRange_lin(X, y_min=0, y_max=1,
+    # Munich when comparing it to Hamburg (changes along the meridian lines), the difference was around 0.1%.
+
+    # joining thw two geodataframe using sjoin, in order to join the dataframe that has the values with its corresponding polygons
+    print("Joining the two geodataframes")
+    join_within_right_gdf_utm32N = gdf_data_utm32N.sjoin(gdf_polygons_utm32N, how="right", predicate="within")
+    
+    shapefile_gdf = gpd.read_file(shapefile)
+    print(shapefile_gdf.crs)   # to check wether it has the same crs as the geodataframe that we created, if not convert it to utm (to be able to calculate the area)
+
+    print("intersecting")
+    overlay_gdf = gpd.overlay(join_within_right_gdf_utm32N, shapefile_gdf, how='intersection')
+    overlay_gdf['area_intersected'] = overlay_gdf.area
+    overlay_gdf['weight'] = overlay_gdf['area_intersected']/overlay_gdf['area'] # this weight represent how much area is inside the shapefile after intesection
+
+
+    # after this step we should have a geodataframe for the intersected polygons that are inside the shapefile of interest
+    # the geodataframe will have the lon,lat, with the corresponding values for each date (each date in a column), 
+    # area of the polygon before intersection, area after intersection,
+    # and the ratio (area_intersected/area) and it can also include the names of the 
+    # regions (in order to be able to dissolve it into mean values for each region)
+
+    # in the next step, a dataframe was created to add the values of each polygon (gridpoint) after calculating
+    # its weight inside the intersection for each date
+    nr_yr = nr_yr
+    nr_entries = nr_entries # ex: number of dates :hourly, daily, monthly..etc
+    nr_column = nr_yr*nr_entries
+    df_data_inter = overlay_gdf.iloc[:,3:nr_column+2].multiply(overlay_gdf['weight'], axis="index")
+
+    # copy the geometry column (important for creating a geodataframe) and other relevant data to the new dataframe
+    # like the name of the regions (important for calculating the mean values for each region)
+    Name_area = 'insert name of the area of interest'
+    df_data_inter[Name_area] = overlay_gdf[Name_area] # this is only an example and should be adapted according to the information in the shapefile
+    df_data_inter['geometry'] = overlay_gdf['geometry']
+    geometry = df_data_inter['geometry']
+    gdf_data_inter = gpd.GeoDataFrame(df_data_inter, crs="epsg:25832", geometry=geometry)
+
+    print("calculating the mean value for each region of interest")
+    dissolve_anomaly_gdf = gdf_data.dissolve(by=Name_area, aggfunc='mean', as_index=True, level=None, sort=True, observed=False, dropna=True)
+    dissolve_anomaly_gdf_wgs = dissolve_anomaly_gdf.to_crs('EPSG:4326') # convert back to wgs 1984
+    
+    print("saving as a shapefile")
+    dissolve_anomaly_gdf_wgs.to_file(save_dir)
+
```

Comment:
Within the doc-string you are writing that there is a return value:
```
    Returns
    ---------
    a shapefile of the area of interest including its corresponding spatial mean values
```
But there is no return statement, you are writing to a file.   
Please decide for one or the other.